src/main/java/org/stellar/sdk/Sep10Challenge.java

```diff
@@ -40,4 +44,336 @@ public static String newChallenge(
+    if (clientSigners.isEmpty()) {
+      throw new InvalidSep10ChallengeException("No verifiable signers provided, at least one G... address must be provided.");
+    }
+
+    // Verify all the transaction's signers (server and client) in one
+    // hit. We do this in one hit here even though the server signature was
+    // checked in the readChallengeTx to ensure that every signature and signer
+    // are consumed only once on the transaction.
+    Set<String> allSigners = new HashSet<String>(clientSigners);
+    allSigners.add(serverKeyPair.getAccountId());
+    LinkedHashSet<String> allSignersFound = verifyTransactionSignatures(transaction, allSigners);
+
+    // Confirm the server is in the list of signers found and remove it.
+    boolean serverSignerFound = allSignersFound.remove(serverKeyPair.getAccountId());
+    // After removing the server signer we call it clientSignersFound
+    LinkedHashSet<String> clientSignersFound = allSignersFound;
+    // Confirm we matched a signature to the server signer.
+    if (!serverSignerFound) {
+      throw new InvalidSep10ChallengeException(String.format("Transaction not signed by server: %s.", serverAccountId));
+    }
+
+    // Confirm we matched signatures to the client signers.
+    if (clientSignersFound.isEmpty()) {
+      throw new InvalidSep10ChallengeException("Transaction not signed by any client signer.");
+    }
+
+    // Confirm all signatures were consumed by a signer.
+    if (clientSignersFound.size() != transaction.getSignatures().size() - 1) {
+      throw new InvalidSep10ChallengeException("Transaction has unrecognized signatures.");
+    }
+
+    return clientSignersFound;
+  }
+
+  /**
+   * Verifies that for a SEP-0010 challenge transaction
+   * all signatures on the transaction are accounted for and that the signatures
+   * meet a threshold on an account. A transaction is verified if it is signed by
+   * the server account, and all other signatures match a signer that has been
+   * provided as an argument, and those signatures meet a threshold on the
+   * account.
+   *
+   * @param challengeXdr    SEP-0010 transaction challenge transaction in base64.
+   * @param serverAccountId Account ID for server's account.
+   * @param network         The network to connect to for verifying and retrieving.
+   * @param threshold       The threshold on the client account.
+   * @param signers         The signers of client account.
+   * @return a list of signers that were found is returned, excluding the server account ID.
+   * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
+   * @throws IOException                    If read XDR string fails, the exception will be thrown.
+   */
+  public static LinkedHashSet<String> verifyChallengeTransactionThreshold(String challengeXdr, String serverAccountId, Network network, int threshold, Set<Signer> signers) throws InvalidSep10ChallengeException, IOException {
```

Comment:
What's the motivation for returning a specific type of Set here? Can we use the generic Set interface?

Reply:
In my opinion, `SignersFound` is in order. This order is the order in which the signers sign. When we return `LinkedHashSet`, the user can explicitly determine that it is in order. 

If I have something wrong please correct me. Thank you.

Reply:
I'm sorry, I got it wrong, I have fixed it.

Reply:
But I think this is worth mentioning, should we return an ordered `signersFound`? This order corresponds to the signatures in the transaction.